### DISCUSSION TITLE
How to ensure that prev provider can be automatically cleared when using keepAlive

### DISCUSSION THREAD

**Discussion by syssam** [rrousselGit/riverpod#3391] [Q&A]
```
@Riverpod(keepAlive: true)
class TodoList extends _$TodoList {
  @override
  List<String> build(String filter) {
    return [
      'A',
      'B',
      'C'
    ];
  }
}
```

From the document, one state per parameter combination will be created, which can lead to memory leaks.
How to ensure that when creating a new provider, the previous provider is automatically cleared

**Comment by idispatch75** [comment]
By using `keepAlive: false`? (or just `@riverpod`).

When using `keepAlive: true` you're basically saying "I want this provider to never be disposed".
There is no "previous" provider here: you create a new provider for each `filter` value, and each one lives forever because you keep it alive.

**Comment by syssam** [comment]
If using keepAlive: false, then the result would not keep when moving to another page
if using keepAlive: true, one state per parameter combination will be created, which can lead to memory leaks
so, how can I create new parameter combination, it will auto dispose last one

**Comment by idispatch75** [comment]
If you want only last result of your filter, then I would remove the parameter in `build` and add a method `void applyFilter(String filter)` that updates the state of the provider.
```dart
@Riverpod(keepAlive: true)
class TodoList extends _$TodoList {
  String _filter = '';

  @override
  List<String> build() {
    final allTodos = ref.watch(todosProvider);

    return _filterList(allTodos);
  }

  List<String> _filterList(List<Todo> todos) {
    // filter todos using _filter
  }

  void applyFilter(String filter) {
    _filter = filter;

    final todos = ref.read(todosProvider);
    state = _filterList(todos);

    // or you can just store _filter and ref.invalidateSelf()
}
```
Your pages would watch the last filtered list.
And when the filter changes you would call `ref.read(todoListProvider.notifier).applyFilter(filter)`.
